python/hsfs/engine/spark.py

```diff
@@ -651,23 +651,25 @@ def _deserialize_from_avro(
         dataframe: Union[RDD, DataFrame],
     ):
         """
-        Deserializes 'value' column from binary using avro schema and unpacks it into columns.
+        Deserializes 'value' column from binary using avro schema.
         """
-        decoded_dataframe = dataframe.select(
-            from_avro("value", feature_group._get_encoded_avro_schema()).alias("value")
-        ).select(col("value.*"))
+        decoded_dataframe = dataframe.withColumn(
+            "value", from_avro("value", feature_group._get_encoded_avro_schema())
+        )
 
         """Decodes all complex type features from binary using their avro type as schema."""
-        return decoded_dataframe.select(
-            [
-                field["name"]
-                if field["name"] not in feature_group.get_complex_features()
-                else from_avro(
-                    field["name"], feature_group._get_feature_avro_schema(field["name"])
-                ).alias(field["name"])
-                for field in json.loads(feature_group.avro_schema)["fields"]
-            ]
-        )
+        for field in json.loads(feature_group.avro_schema)["fields"]:
+            name = field["name"]
+            full_field = f"value.{name}"
+
+            if name in feature_group.get_complex_features():
+                decoded_dataframe = decoded_dataframe.withColumn(
+                    full_field, from_avro(col(full_field), feature_group._get_feature_avro_schema(name))
+                )
+            else:
+                decoded_dataframe = decoded_dataframe.withColumn(full_field, col(full_field))
```

Comment:
This is not a good idea. Using `withColumn` within a for loop makes Spark super slow and potentially breaks Spark. 
This is because every `withColumn` adds a new operation in the Spark Tree that the compiler and optimizer need to deal with.

You should assemble the list of columns you need and then do a single Spark operation - (like the code was before).